solr/core/src/java/org/apache/solr/schema/ManagedIndexSchemaFactory.java

```diff
@@ -198,88 +198,94 @@ public ManagedIndexSchema create(
+        } catch (KeeperException.NoNodeException e) {
+          log.info(
+              "The schema is configured as managed, but managed schema resource {} not found - loading non-managed schema {} instead",
+              managedSchemaResourceName,
+              resourceName);
+        } catch (KeeperException e) {
+          String msg = "Error attempting to access " + managedSchemaPath;
+          log.error(msg, e);
+          throw new SolrException(ErrorCode.SERVER_ERROR, msg, e);
+        }
+        if (null == schemaInputStream) {
+          // The managed schema file could not be found - load the non-managed schema
           try {
-            // Retry to load the managed schema, in case it was created since the first attempt
-            byte[] data = zkClient.getData(managedSchemaPath, null, stat, true);
-            schemaZkVersion = stat.getVersion();
-            schemaInputStream = new ByteArrayInputStream(data);
-            loadedResource = managedSchemaPath;
-            warnIfNonManagedSchemaExists();
-          } catch (Exception e1) {
-            if (e1 instanceof InterruptedException) {
-              Thread.currentThread().interrupt(); // Restore the interrupted status
+            schemaInputStream = loader.openResource(resourceName);
+            loadedResource = resourceName;
+            shouldUpgrade = true;
+          } catch (IOException e) {
+            try {
+              // Retry to load the managed schema, in case it was created since the first attempt
+              byte[] data = zkClient.getData(managedSchemaPath, null, stat, true);
+              schemaZkVersion = stat.getVersion();
+              schemaInputStream = new ByteArrayInputStream(data);
+              loadedResource = managedSchemaPath;
+              warnIfNonManagedSchemaExists();
+            } catch (Exception e1) {
+              if (e1 instanceof InterruptedException) {
+                Thread.currentThread().interrupt(); // Restore the interrupted status
+              }
+              final String msg =
+                  "Error loading both non-managed schema '"
+                      + resourceName
+                      + "' and managed schema '"
+                      + managedSchemaResourceName
+                      + "'";
+              log.error(msg, e);
+              throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, msg, e);
             }
-            final String msg =
-                "Error loading both non-managed schema '"
-                    + resourceName
-                    + "' and managed schema '"
-                    + managedSchemaResourceName
-                    + "'";
-            log.error(msg, e);
-            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, msg, e);
           }
         }
       }
+      InputSource inputSource = new InputSource(schemaInputStream);
+      inputSource.setSystemId(SystemIdResolver.createSystemIdFromResourceName(loadedResource));
```

Comment:
I don't think a SchemaFactory needs to be thread-safe.  I see that they are always created in order to call create() and then are thrown away, not saved for potential re-use.  Ideally if we annotate this class with `net.jcip.annotations.NotThreadSafe` one would hope tooling could be smarter to understand that THREAD_SAFETY_VIOLATION is impossible here.  Lift doesn't do its own analysis, it delegates to tools like "Infer" which has a module "RacerD" that implements this particular THREAD_SAFETY_VIOLATION analysis.  I was looking into this last week -- it's kind of fascinating.  I want to help RacerD help Lift help us.  When I look at RacerD docs, it sadly seems a bit dated; doesn't recognized popular annotations like NotThreadSafe; it has its own.  In this case, an annotation [ThreadConfined](https://fbinfer.com/docs/1.0.0/checker-racerd/#threadconfined) should be equivalent.  I'll report an issue there and to Lift as well.

Reply:
Thanks for the reassurance and extra explanation!

3a047d404b4e994865711229329a7461a751897b should fix this. Whether or not this class needs to be thread-safe, it seems weird to me that `loadedResource` is an instance variable -- seems mostly to exist in order to pass extra info from the `readSchemaLocally()` method back to the `create()` method, which could be straightforwardly done another way. I'm happy to include 3a047d404b4e994865711229329a7461a751897b in this PR or revert it from the PR branch, depending on what you prefer.

Reply:
A dev there clarified that NotThreadSafe should be understood!  I think I'd prefer that to 3a047d4 which feels awkward because the Array wrapping for modifying loadedResource.

Reply:
To be clear, I'm really okay with that change.  You've added it I see; it's fine.

Reply:
You're ok with 3a047d404b4e994865711229329a7461a751897b ? The array wrapping is admittedly awkward, but IMO much clearer than (and preferable to) setting an instance variable in one method and then checking its value in another, which is the status quo.

Reply:
Yeah it's fine.